pyproject.toml

```diff
@@ -110,3 +110,100 @@ source-includes = [
 [build-system]
 requires = ["pdm-backend"]
 build-backend = "pdm.backend"
+
+[tool.ruff.lint]
+# see https://docs.astral.sh/ruff/configuration/#using-pyprojecttoml
+# "F" contains autoflake, see https://github.com/astral-sh/ruff/issues/1647
+# add more rules
+select = [
+    # default Ruff checkers as of ruff 0.1.3: E4, E7, E9, F
+    "E4",
+    "E7",
+    "E9",
+    "F",  # pyflakes
+
+    # the rest in alphabetical order:
+    # TODO: "A",   # flake8-builtins
+    # TODO: "ARG", # flake8-unused-arguments
+    "B",   # flake8-bugbear
+    # TODO: "BLE", # flake8-blind-except
+    # TODO: Do I want "COM", # flake8-commas
+    "C4",  # flake8-comprehensions
+    # TODO: "DJ",  # flake8-django
+    # TODO: "DTZ", # flake8-datetimez
+    # TODO: "EM",  # flake8-errmsg
+    "EXE", # flake8-executable
+    # TODO: "FURB", # refurb
+    # TODO: "FBT", # flake8-boolean-trap
+    # TODO: "G",   # flake8-logging-format
+    "I",   # isort
+    "ICN", # flake8-import-conventions
+    "INP", # flake8-no-pep420
+    # TODO: "INT", # flake8-gettext
+    "ISC", # flake8-implicit-str-concat
+    # TODO: "LOG", # flake8-logging
+    "PERF", # perflint
+    "PIE", # flake8-pie
+    "PL",  # pylint
+    "PYI", # flake8-pyi
+    # TODO: "RET", # flake8-return
+    "RSE", # flake8-raise
+    "RUF",
+    # TODO: "SIM", # flake8-simplify
+    "SLF", # flake8-self
+    "SLOT", # flake8-slots
+    "TID", # flake8-tidy-imports
+    "UP",  # pyupgrade
+    "Q",   # flake8-quotes
+    "TCH", # flake8-type-checking
+    "T10", # flake8-debugger
+    "T20", # flake8-print
+    # TODO: "S",   # flake8-bandit
+    "YTT", # flake8-2020
+    # TODO: add more flake8 rules
+  ]
+
+  ignore = [
+    # suppression in order of # of violations:
+    "B007",
+    "T201",
+    "PLW2901",
+    "SLF001",
+    "RUF001",
+    "PLR2004",
+    "PLR0912",
+    "PLR0913",
+    "RUF005",
+    "RUF012",
+    "PLR0915",
+    "INP001",
```

Comment:
I don't get this. `INP` has only one rule `INP001`. What's the point of adding checks for it but ignoring the only thing it checks. I didn't look at all of them but there are other similar ones. e.g. `ISC` has 3 rules but two off them are ignored.

Reply:
I just did the quick thing: added each thing in the list, suppressed those subrules with violations, figuring there were probably other rules being checked. It's like closing the gate, so we don't introduce new violations.

You've found a couple of weird cases, it's true, but I don't understand how you'd like to move forward.

I thought it was probably useful to enforce a bunch of the rules quickly. Perhaps you disagree, but what now?

Reply:
I just want to understand the logic behind some checks. 

* `INP` for example disallows implicit namespace packages. That's not going to work. We have namespace packages and will probably add more. Why should we disallow that? I understand the reason behind the rule: "Hey, maybe you forgot to add `__init__.py`?" but that's not applicable here. We intentionally use implicit namespace packages.
* `T20` disallows `print` (or its cousin `pprint`), which is a very project specific convention. We clearly have scripts that `print` stuff (`pelican-quickstart` etc). Either we don't care about this rule or we switch from `print` to something else. Until then, this rule is effectively not enforceable.
* `ISC` is one thing I can agree with but we are ignoring most of them :).

I understand that to un-ignore some of these rules, perhaps some manual fixes are needed. That is fine but then I'd treat this as a "work-in-progress" and the ignored rules more like a "to-do list" :).

Reply:
I really appreciate your careful review. Thanks.

My overall goal is to make some progress here, merge something. In particular, the auto-fix ("--fix") and auto-format.

I agree the ignored rules are like a TODO list. I was hoping to merge that TODO list, since it represents some knowledge that might be useful to lay out in the codebase.

However, if you don't want to merge that, then another possibility is to take out the extra rules and the suppressions.

What is the quickest path to moving forward?

Reply:
Another possibility would be to not include any rules that have suppressions, and drop INP and T20. Not sure where that would leave the list, I'd have to try it.

Reply:
At least a comment next to the ignored rule stating what it does would be useful. It would have saved me some work during review and I'd guess it'll be useful when someone picks up these rules to make them pass. I already started doing that, might was well put it here to make your job easier :).

```python
ignore = [
    # suppression in order of # of violations:
  "B007",  # unused-loop-control-variable
  "T201",  # print
  "PLW2901",  # redefined-loop-name
  "SLF001",  # private-member-access
  "RUF001",  # ambiguous-unicode-character-string
  "PLR2004",  # magic-value-comparison
  "PLR0912",  # too-many-branches
  "PLR0913",  # too-many-arguments
  "RUF005",  # collection-literal-concatenation
  "RUF012",  # mutable-class-default
  "PLR0915",  # too-many-statements
  "INP001",  # implicit-namespace-package
  "RUF015",  # unnecessary-iterable-allocation-for-first-element
  "PLR1722",  # sys-exit-alias
  "ISC001",  # single-line-implicit-string-concatenation
  "C408",  # unnecessary-collection-call
  "B904",  # raise-without-from-inside-except
  "UP031",  # printf-string-formatting
  "PLR5501",  # collapsible-else-if
  "PERF203",  # try-except-in-loop
  "B006",  # mutable-argument-default
  "PLR1714",  # repeated-equality-comparison
  "PERF401",  # manual-list-comprehension
  # TODO: these only have one violation each:
  "SLOT000",  # no-slots-in-str-subclass
  "PYI024",  # collections-named-tuple
  "PLW0603",  # global-statement
  "PIE800",  # unnecessary-spread
  "ISC003",  # explicit-string-concatenation
  "EXE002",  # shebang-missing-executable-file
  "C401",  # unnecessary-generator-set
  "B028",  # no-explicit-stacklevel
  "B008",  # function-call-in-default-argument
]
```

Reply:
Also probably drop these comments:
> suppression in order of # of violations:
> TODO: these only have one violation each:

While they might be accurate at the time, these kind of comments have a very strong tendency to become incorrect and/or misleading quickly.

Reply:
Good idea. Done.

Reply:
"Done" for adding the comments.

For dropping, how about I add "in Dec 2023"? That would preserve the info (these were in order) without lying.

If I were to work on the TODO list, I'd start with the ones that had one violation each in 2023, it might be helpful to have them marked.

Reply:
I added the "in Dec 2023". But I'm happy to drop those comments entirely if you'd prefer.

Reply:
I agree that comments next to rules are helpful. As I've mentioned before, the set of rules and ignores that I've used for some of my plugin code-style check experimentation is an example of that practice: https://github.com/pelican-plugins/featured-image/blob/3d3473269c27dfd3ae123a73aebeb1af4d8b3dd0/pyproject.toml#L58-L91